CLI specification

nikololiahim opened the pull request.

deemp commented on docs/cli-spec.md:
```diff
@@ -0,0 +1,40 @@
+* The subcommand specifies which files should be analyzed (`.eo`, `.java` or `.py`). More languages can be added in the future. 
+* Analyzes the input files in the `--in` directory. If `--in` is not specified, defaults to reading the code in the input language from stdin. 
+* `--in` can also accept a path which leads to a file. In this case, only the specified file will be analyzed. 
+* The temporary files produced by analyzers are available in `--tmp` directory.  If `--tmp` is not specified, temporary files are stored in the OS-created tempdir. Each target language may have a different structure of the files in the temporary directory.
```
1. This implies `[--tmp <path>]`. 
I suppose, the flag should be `[--tmp <path>]` and the definition of `<thing>` should be added. Or is this a convention?

nikololiahim replied:
@br4ch1st0chr0n3 I rewrote the description using [this](https://en.wikipedia.org/wiki/Command-line_interface#:~:text=Command%20description%20syntax%5Bedit%5D) syntax guide.  What do you think?

deemp commented on docs/cli-spec.md:
```diff
@@ -0,0 +1,40 @@
+
+## Output configuration
+* `--sarif` option means that the output of the analysis will be in the [SARIF ](https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html) format. If this option is not specified only the plain text console output will be produced.
+* `--files [<path>]` option specifies whether the output should be written to the output directory instead of writing it to the console. 
```
probably need `[--files <path>]`

deemp commented on docs/cli-spec.md:
```diff
@@ -0,0 +1,40 @@
+## Output configuration
+* `--sarif` option means that the output of the analysis will be in the [SARIF ](https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html) format. If this option is not specified only the plain text console output will be produced.
+* `--files [<path>]` option specifies whether the output should be written to the output directory instead of writing it to the console. 
+If the path is specified, write the files to the given path. If not, then creates a directory called `polystat_out` and write the files there. 
```
If it's `./polystat_out`, need to add it to `.gitignore'. Some projects I touched recently didn't do that with their respective outputs. Just a complaint

deemp commented on docs/cli-spec.md:
```diff
@@ -0,0 +1,40 @@
+* `--sarif` option means that the output of the analysis will be in the [SARIF ](https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html) format. If this option is not specified only the plain text console output will be produced.
+* `--files [<path>]` option specifies whether the output should be written to the output directory instead of writing it to the console. 
+If the path is specified, write the files to the given path. If not, then creates a directory called `polystat_out` and write the files there. 
+if a non-default output format is specified, e.g. `--sarif`, then the files created by the analysis should be written in the respective subdirectory. For example, in case of `--sarif`,  the SARIF files will be located in `<path>/sarif/`. The console output is not written anywhere. Therefore, if none of the output format options (e.g. `--sarif`) are specified, no files are produced. 
```
`created by the analysis` -> `... analyzer`?

deemp commented on docs/cli-spec.md:
```diff
@@ -0,0 +1,40 @@
+* `--files [<path>]` option specifies whether the output should be written to the output directory instead of writing it to the console. 
+If the path is specified, write the files to the given path. If not, then creates a directory called `polystat_out` and write the files there. 
+if a non-default output format is specified, e.g. `--sarif`, then the files created by the analysis should be written in the respective subdirectory. For example, in case of `--sarif`,  the SARIF files will be located in `<path>/sarif/`. The console output is not written anywhere. Therefore, if none of the output format options (e.g. `--sarif`) are specified, no files are produced. 
+If `--in` option specifies a directory, the structure of the output files should be similar to the structure of the input directory. 
```
output files should be similar -> ... files will be similar

deemp commented on docs/cli-spec.md:
````diff
@@ -0,0 +1,40 @@
+polystat java --in src/main/java
+```
+
+2. Write the SARIF JSON files to `polystat_out/sarif` from analysing the 'tmp' directory with `.eo` files.
````
`tmp` with backticks

deemp commented on docs/cli-spec.md:
````diff
@@ -0,0 +1,40 @@
+polystat (eo | java | python) --tmp <path> --in <path> [--include <rule>... | --exlude <rule>...] --sarif  --files [<path>]
+polystat --version --help
+polystat list
+```
````
I think we need to add a `--config <file>` flag here or define `polystat with <path>`, where `path` leads to a config file

nikololiahim replied:
I thought about this. It would probably be best to employ some configuration format and make it so that it's possible to specify only certain options. We can use [ciris](https://cir.is/) to populate a Config case class from some of the supported formats. If the same option is specified both in the config and in the cmdline, cmdline option should take precedence.

nikololiahim commented:
@br4ch1st0chr0n3 I rewrote the description using [this](https://en.wikipedia.org/wiki/Command-line_interface#:~:text=Command%20description%20syntax%5Bedit%5D) syntax guide.  What do you think?

deemp commented on docs/cli-spec.md:
```diff
@@ -1,29 +1,27 @@
-polystat (eo | java | python) --tmp <path> --in <path> [--include <rule>... | --exlude <rule>...] --sarif  --files [<path>]
-polystat --version --help
+polystat {eo | java | python} [--tmp <path>] [--in <path>] [{--include <rule...> | --exclude <rule...>}] [--sarif] [--files [path]]
+polystat [--version] [--help]
```
What's will be the output of calling simply `polystat`?

nikololiahim replied:
The same as `polystat --help`, I think. Can add this to the description.

deemp commented on docs/cli-spec.md:
```diff
@@ -1,29 +1,27 @@
-`(...)` - must be present
-`[...]` - may not be present
-`<thing>...` -- multiple instances of thing can be accepted.
+The description follows [this guide](https://en.wikipedia.org/wiki/Command-line_interface#:~:text=Command%20description%20syntax%5Bedit%5D).
```
This link refers to the relevant section in the article ([link](https://en.wikipedia.org/wiki/Command-line_interface#Command_description_syntax)). Taken from the *Contents* session from the top of the page.

nikololiahim replied:
Ok, will update.

deemp commented on docs/cli-spec.md:
````diff
@@ -1,29 +1,27 @@
 ```
-polystat (eo | java | python) --tmp <path> --in <path> [--include <rule>... | --exlude <rule>...] --sarif  --files [<path>]
-polystat --version --help
+polystat {eo | java | python} [--tmp <path>] [--in <path>] [{--include <rule...> | --exclude <rule...>}] [--sarif] [--files [path]]
````
How about adding a `polystat config <file>` option? This will read the configuration from a file and won't interfere with the terminal-supplied commands. Or, config will be overriden by the terminal-supplied commands. What do you think?

nikololiahim replied:
Hmm, this can actually be a good use case for the plain `polystat` command. I like this idea! It will try to discover a config file in the current directory and read the options from there.

nikololiahim replied:
If some alternative location needs to be specified, `polystat --config <path>` can be used instead.

deemp commented on docs/cli-spec.md:
````diff
@@ -1,29 +1,27 @@
 ```
-polystat (eo | java | python) --tmp <path> --in <path> [--include <rule>... | --exlude <rule>...] --sarif  --files [<path>]
-polystat --version --help
+polystat {eo | java | python} [--tmp <path>] [--in <path>] [{--include <rule...> | --exclude <rule...>}] [--sarif] [--files [path]]
````
This description of CLI is much clearer, thank you!

nikololiahim commented:
@br4ch1st0chr0n3 I think I have addressed all the suggestions in https://github.com/nikololiahim/polystat-cli/pull/4/commits/e3a4a6c4c2f1d9fe9ab06520b5793dfc3a262617. What do you think?

nikololiahim commented:
The only undecided thing for now is the format of the configuration file. I suggest using the ones supported by [ciris](https://cir.is/), beacause they would be the easiest to integrate with. This means we have [JSON](https://cir.is/docs/modules#circe), [YAML](https://cir.is/docs/modules#circe-yaml) and [HOCON](https://github.com/2m/ciris-hocon) to choose from.

nikololiahim commented:
Maybe the language shouldn't be mutually exclusive in the end. It is possible to differentiate between the generated files if we also include the original extension in the output file name, e.g. `hello.eo.sarif.json`, `hello.java.sarif.json`, etc.

nikololiahim commented:
But before that, it's important to finish reading configs from the file. I think I'll settle on HOCON with ciris. We'll see how it goes though.

br4ch1st0chr0n3 commented:
> But before that, it's important to finish reading configs from the file. I think I'll settle on HOCON with ciris. We'll see how it goes though.

I am unsure that we need to use a special format of config file [1] and not a plain text file [2]. 
Benefits of [1]: syntax checking and highlighting, allows to split long commands
Drawbacks of [1]: unnecesarily complex, need a separate section describing how to write such file

Benefits of [2]: has the same format as specified here (hence, can be directly copied into a console), allows splitting long commands, is easy to read
```sh
polystat
java
--in in
```
Drawbacks of [2]: no syntax checking and highlighting

Both options can be used. However, the commands won't probably be too long, so I'd prefer using a simpler option. 
But I don't mind if you'd like to experiment with a new config format.

nikololiahim commented:
@br4ch1st0chr0n3 some comments are addressed in https://github.com/nikololiahim/polystat-cli/pull/4/commits/35681d01c8063667f9daaf223caa4df80d911bce
